torch/_inductor/kernel/mm_scaled.py

```diff
@@ -17,12 +19,204 @@
     TritonTemplate,
 )
 from ..utils import use_aten_gemm_kernels, use_ck_gemm_template, use_triton_template
-from .mm_common import _is_static_problem, mm_args, mm_grid, scaled_mm_configs
+from .mm_common import (
+    _is_static_problem,
+    mm_args,
+    mm_grid,
+    persistent_grid,
+    persistent_mm_configs,
+    scaled_mm_configs,
+)
 
 
 log = logging.getLogger(__name__)
 aten = torch.ops.aten
 
+load_scales = r"""
+@triton.jit
+def load_scales(a_scale_ptr, b_scale_ptr, SCALING_ROWWISE: tl.constexpr):
+    if SCALING_ROWWISE:
+        # For row-wise scaling, we'll return the pointers
+        return a_scale_ptr, b_scale_ptr
+    else:
+        # For per-tensor scaling, we'll load the scalar values
+        a_scale = tl.load(a_scale_ptr)
+        b_scale = tl.load(b_scale_ptr)
+        return a_scale, b_scale
+"""
+
+
+apply_scaling = r"""
+@triton.jit
+def apply_scaling(
+    accumulator,
+    a_scale,
+    b_scale,
+    SCALING_ROWWISE: tl.constexpr,
+    offs_cm,
+    offs_cn,
+    M,
+    N,
+    stride_a_scale_m,
+    stride_b_scale_n,
+):
+    if SCALING_ROWWISE:
+        # For row-wise scaling, we need to load the scales for each row/column
+        a_scales = tl.load(
+            a_scale + (offs_cm * stride_a_scale_m),
+            mask=offs_cm < M,
+            other=0.0,
+        )
+        b_scales = tl.load(
+            b_scale + (offs_cn * stride_b_scale_n),
+            mask=offs_cn < N,
+            other=0.0,
+        )
+        acc_scale = a_scales[:, None] * b_scales[None, :]
```

Comment:
The exception text below says "or 1-dimensional tensors with the same size". Here seems that they are M- and N-sized. Is it always that M == N?

Reply:
Hmm so that is just a bad error message, its really saying that scale_a.ndim() ==  scale_b.ndim(). This check is a little weird and is supposed to reflect:
https://github.com/pytorch/pytorch/blob/9039fbb47ecfc93df74a014a209e5929d10fd2a3/aten/src/ATen/native/cuda/Blas.cpp#L902